pkg/image/prune/imagepruner.go

```diff
@@ -488,10 +557,33 @@ func (p *imagePruner) Run(pruneImage ImagePruneFunc, pruneStream ImageStreamPrun
 	allNodes := p.g.NodeList()
 
 	imageNodes := imageNodeSubgraph(allNodes)
+	if len(imageNodes) == 0 {
+		return
+	}
+
+	// we only support a single internal registry, and all images should have the same registry
+	// so we just take the 1st one and use it
+	ref, err := imageapi.ParseDockerImageReference(imageNodes[0].Image.DockerImageReference)
+	if err != nil {
+		glog.Errorf("error determining registry URL from %q: %v", imageNodes[0].Image.DockerImageReference, err)
+		return
+	}
+
+	registryURL := ref.Registry
+	if len(registryURL) == 0 {
+		glog.Errorf("unable to prune - registry URL is missing")
+		return
+	}
+
+	if err := p.registryPinger.ping(registryURL); err != nil {
+		glog.Errorf("error communicating with registry: %v", err)
+		return
+	}
+
 	pruneImages(p.g, imageNodes, pruneImage, pruneStream, pruneManifest)
 
 	layerNodes := layerNodeSubgraph(allNodes)
-	pruneLayers(p.g, layerNodes, pruneLayer, pruneBlob)
+	pruneLayers(p.g, registryURL, layerNodes, pruneLayer, pruneBlob)
```

Comment:
I 100% agree. That's what #3333 is about. However, this PR should eliminate certain problems (it rejects pruning if the registry isn't reachable, it prunes blobs if there are no streams referencing them) but it doesn't address the overall issue. I do not believe this PR makes things worse.

I can modify it so pruneLayers is called first, and if any prune-layer or prune-blob call fails, don't call pruneImages. That would at least avoid deleting data from etcd that we'd need to use to determine the layer/blob deletions to retry. Tombstoning would probably need to come as part of the bigger refactor.